CRAYSAT-1957: Functional tests for `sat firmware`

annapoorna-s-alt opened the pull request:
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

_Create functional tests for sat firmware that run as part of the tests in csm-testing_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CRAYSAT-1957](https://jira-pro.it.hpe.com:8443/browse/CRAYSAT-1957)

## Testing

_List the environments in which these changes were tested._

### Tested on:

  * fanta

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

Tested with goss test downloading unstable rpm

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_
NO

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

annapoorna-s-alt commented:
OUTPUT

```
ncn-m001:/mnt/developer/sann/firmware-AT # ./opt/cray/tests/install/csm-testing-python-venv/bin/test_sat_functional firmware.test_firmware
test_firmware_creates_and_deletes_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that running `sat firmware` with no options creates a snapshot and outputs a report ... ok
test_firmware_describe_existing_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots SNAPSHOT` describes an existing firmware snapshot. ... ok
test_firmware_list_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots` prints a list of snapshot names. ... ok
test_firmware_query_multiple_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots SNAPSHOT_1 SNAPSHOT_2` prints the details for two snapshots. ... ok
test_firmware_query_multiple_xnames (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware -x <xname1>,<xname2>` prints firmware info for multiple xnames. ... ok
test_firmware_query_nonexistent_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots NON_EXISTENT_SNAPSHOT` querying a non-existent firmware snapshot. ... ok
test_firmware_query_single_xname (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware -x <xname>` prints firmware info for a single xname. ... ok
test_firmware_query_xnames_from_file (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --xname-file <file>` prints firmware info for xnames listed in a file. ... ok

----------------------------------------------------------------------
Ran 8 tests in 146.911s

OK

```

ethanholen-hpe commented:
Hey Annapoorna, I tested this on mug and it looks like it is failing on one of the tests but I cant tell if that's a problem with mug or with the test. Ill give this an approval anyway since the rest looks great to me. Heres the output from mug:

```
ncn-m001:/mnt/developer/eholen # printf "\nname:\n $(echo $SYSTEM_NAME)\n\nSAT Version:\n $(sat --version)\n\n" && printf "System Version(s):\n" && kubectl get cm -n services cray-product-catalog -o json | jq  -r '.data.csm' | yq r -  -d '*' -j | jq -r 'keys[]'

name:
 mug

SAT Version:
 sat 3.35.15

System Version(s):
1.5
1.6.0-alpha.58
1.6.1-casmcms-9262
1.7
ncn-m001:/mnt/developer/eholen # ./opt/cray/tests/install/csm-testing-python-venv/bin/test_sat_functional firmware.test_firmware
test_firmware_creates_and_deletes_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that running `sat firmware` creates a snapshot and outputs a report ... ok
test_firmware_describe_existing_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots SNAPSHOT` describes an existing firmware snapshot. ... ok
test_firmware_list_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots` prints a list of snapshot names. ... FAIL
test_firmware_query_multiple_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots SNAPSHOT_1 SNAPSHOT_2` prints the details for two snapshots. ... ok
test_firmware_query_multiple_xnames (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware -x <xname1>,<xname2>` prints firmware info for multiple xnames. ... ok
test_firmware_query_nonexistent_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots NON_EXISTENT_SNAPSHOT` querying a non-existent firmware snapshot. ... ok
test_firmware_query_single_xname (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware -x <xname>` prints firmware info for a single xname. ... ok
test_firmware_query_xnames_from_file (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --xname-file <file>` prints firmware info for xnames listed in a file. ... ok

======================================================================
FAIL: test_firmware_list_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots` prints a list of snapshot names.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/developer/eholen/opt/cray/tests/install/csm-testing-python-venv/lib64/python3.6/site-packages/csm_testing/tests/sat_functional/firmware/test_firmware.py", line 184, in test_firmware_list_snapshots
    self.assertRegex(line, pattern, f"Line does not look like a snapshot name: {line}")
AssertionError: Regex didn't match: '^SAT-\\d{4}-\\d{1,2}-\\d{1,2}-\\d{1,2}-\\d{1,2}-\\d{1,2}$' not found in 'fullSystem_joshuaw_20250107.1' : Line does not look like a snapshot name: fullSystem_joshuaw_20250107.1

----------------------------------------------------------------------
Ran 8 tests in 139.301s

FAILED (failures=1)
```

annapoorna-s-alt commented:
> Hey Annapoorna, I tested this on mug and it looks like it is failing on one of the tests but I cant tell if that's a problem with mug or with the test. Ill give this an approval anyway since the rest looks great to me. Heres the output from mug:
> 
> ```
> ncn-m001:/mnt/developer/eholen # printf "\nname:\n $(echo $SYSTEM_NAME)\n\nSAT Version:\n $(sat --version)\n\n" && printf "System Version(s):\n" && kubectl get cm -n services cray-product-catalog -o json | jq  -r '.data.csm' | yq r -  -d '*' -j | jq -r 'keys[]'
> 
> name:
>  mug
> 
> SAT Version:
>  sat 3.35.15
> 
> System Version(s):
> 1.5
> 1.6.0-alpha.58
> 1.6.1-casmcms-9262
> 1.7
> ncn-m001:/mnt/developer/eholen # ./opt/cray/tests/install/csm-testing-python-venv/bin/test_sat_functional firmware.test_firmware
> test_firmware_creates_and_deletes_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that running `sat firmware` creates a snapshot and outputs a report ... ok
> test_firmware_describe_existing_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware --snapshots SNAPSHOT` describes an existing firmware snapshot. ... ok
> test_firmware_list_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware --snapshots` prints a list of snapshot names. ... FAIL
> test_firmware_query_multiple_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware --snapshots SNAPSHOT_1 SNAPSHOT_2` prints the details for two snapshots. ... ok
> test_firmware_query_multiple_xnames (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware -x <xname1>,<xname2>` prints firmware info for multiple xnames. ... ok
> test_firmware_query_nonexistent_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware --snapshots NON_EXISTENT_SNAPSHOT` querying a non-existent firmware snapshot. ... ok
> test_firmware_query_single_xname (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware -x <xname>` prints firmware info for a single xname. ... ok
> test_firmware_query_xnames_from_file (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware --xname-file <file>` prints firmware info for xnames listed in a file. ... ok
> 
> ======================================================================
> FAIL: test_firmware_list_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
> Test that `sat firmware --snapshots` prints a list of snapshot names.
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/mnt/developer/eholen/opt/cray/tests/install/csm-testing-python-venv/lib64/python3.6/site-packages/csm_testing/tests/sat_functional/firmware/test_firmware.py", line 184, in test_firmware_list_snapshots
>     self.assertRegex(line, pattern, f"Line does not look like a snapshot name: {line}")
> AssertionError: Regex didn't match: '^SAT-\\d{4}-\\d{1,2}-\\d{1,2}-\\d{1,2}-\\d{1,2}-\\d{1,2}$' not found in 'fullSystem_joshuaw_20250107.1' : Line does not look like a snapshot name: fullSystem_joshuaw_20250107.1
> 
> ----------------------------------------------------------------------
> Ran 8 tests in 139.301s
> 
> FAILED (failures=1)
> ```



Hey Ethan,
It looks like someone wantedly changed inside code to get their own snapshot name. And they didn't deleted those snapshots after use. So it is not matching with the regex format for those which has different names other than `name = 'SAT-{}-{}-{}-{}-{}-{}'.format( now.year, now.month, now.day, now.hour, now.minute, now.second )`
Here is few snapshots name differed:
```
ncn-m001:~ # sat firmware --snapshots | grep "fullSystem"
fullSystem_joshuaw_20250107.6
fullSystem_joshuaw_20250107.3
fullSystem_joshuaw_20250107.8
fullSystem_joshuaw_20240305
fullSystem_joshuaw_20250107.7
fullSystem_joshuaw_20250107.1
fullSystem_joshuaw_20250107.2
fullSystem_joshuaw_20250107
fullSystem_joshuaw_20250107.10
fullSystem_joshuaw_20250107.5
fullSystem_joshuaw_20250107.9
fullSystem_joshuaw_20250107.4
```

I would request @haasken-hpe also to comment on this

haasken-hpe commented on goss-testing/tests/ncn/goss-sat-firmware.yaml:
```diff
@@ -0,0 +1,41 @@
+            sev: 0
+        exec: |-
+            "{{$logrun}}" -l "{{$test_label}}" \
+                "{{$sat_test}}" "showrev.test_firmware"
```
Incorrect path:

```suggestion
                "{{$sat_test}}" "firmware.test_firmware"
```

haasken-hpe commented:
> Hey Ethan, It looks like someone wantedly changed inside code to get their own snapshot name. And they didn't deleted those snapshots after use. So it is not matching with the regex format for those which has different names other than `name = 'SAT-{}-{}-{}-{}-{}-{}'.format( now.year, now.month, now.day, now.hour, now.minute, now.second )` Here is few snapshots name differed:
> 
> ```
> ncn-m001:~ # sat firmware --snapshots | grep "fullSystem"
> fullSystem_joshuaw_20250107.6
> fullSystem_joshuaw_20250107.3
> fullSystem_joshuaw_20250107.8
> fullSystem_joshuaw_20240305
> fullSystem_joshuaw_20250107.7
> fullSystem_joshuaw_20250107.1
> fullSystem_joshuaw_20250107.2
> fullSystem_joshuaw_20250107
> fullSystem_joshuaw_20250107.10
> fullSystem_joshuaw_20250107.5
> fullSystem_joshuaw_20250107.9
> fullSystem_joshuaw_20250107.4
> ```
> 
> I would request @haasken-hpe also to comment on this

We discussed this on our call today, but I'll just follow up here as well. These snapshots were almost definitely created outside of `sat firmware`, e.g. with the `cray fas` command or with direct calls to the FAS API. In our tests, we cannot assume that such other snapshots do not exist on the system. We should ignore them in our tests.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+import os
+
+SAT_FIRMWARE_HEADER = ['xname', 'name', 'target_name', 'version']
+
```
PEP8 requires another blank line

```suggestion


```

haasken-hpe commented on goss-testing/tests/ncn/goss-sat-firmware.yaml:
```diff
@@ -0,0 +1,41 @@
+        exit-status: 0
+        stderr:
+        - OK
+        timeout: 200000 # timeout in milliseconds
```
I think this may not be enough time on other systems. E.g., here is the timing of just running `sat firmware` on fanta:

```
real	0m17.995s
user	0m0.326s
sys	0m0.281s
```

vs. on rocket:

```
real	1m42.887s
user	0m0.165s
sys	0m0.250s
```

One thing that I think would help would be to only do certain setup tasks only once for all the tests in your `TestFirmware` class, e.g. in a `setUpClass` method. I'll leave comments on that below.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+
+def get_xnames_from_firmware_output() -> list:
+    """Run 'sat firmware' and return a list of unique xnames from the output."""
+    command = "sat firmware"
```
The `sat firmware` command can take quite a while to get a snapshot of the whole system. There may be a better way to get a list of xnames that we would expect to be able to query in FAS.

For example, we might be able to do a query of HSM to get all the NodeBMCs.

From fanta:

```
ncn-m001:/mnt/developer/haasken # cray fas snapshots describe SAT-2025-7-9-21-30-51 | jq -r '.devices[].xname' | sort > fas-xnames.txt
ncn-m001:/mnt/developer/haasken # cray hsm inventory hardware list --type NodeBMC | jq -r '.[].ID' | sort > hsm-xnames.txt
ncn-m001:/mnt/developer/haasken # diff -u fas-xnames.txt hsm-xnames.txt
--- fas-xnames.txt	2025-07-09 21:39:47.459043673 +0000
+++ hsm-xnames.txt	2025-07-09 21:39:56.343867061 +0000
@@ -1,4 +1,3 @@
-x3000c0r26b0
 x3000c0s11b0
 x3000c0s13b0
 x3000c0s15b0
```

So it looks like every NodeBMC xname in HSM was present in FAS, but there was one additional xname present in FAS for some reason. For the purposes of these tests, I think getting a list from `cray hsm inventory list` or a similar hsm command would be pretty reasonable and faster. My understanding is that the creation of a FAS snapshot actually requires FAS to go out and query Redfish endpoints as opposed to just querying the HSM API that looks in a database.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+    return xnames
+
+class TestFirmware(unittest.TestCase):
+
```
Whether you end up improving the performance of `get_xnames_from_firmware_output` by changing it to query HSM or whether you leave it as-is, it would help to only run this one time, so that we are not repeatedly calling FAS to create a whole-system snapshot for every test that needs the list of xnames. Instead, you can utilize the `setUpClass` method of a unittest to only do this one time for all the test methods in this class.

For example:

```
    @classmethod
    def setUpClass(cls):
        cls.xnames = get_xnames_from_firmware_output()
```

Then test methods can access that list of xnames as `self.xnames`.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+            r'INFO: Use `sat firmware --delete-snapshot` to delete the snapshot if it is no longer needed'
+        )
+        match = re.search(pattern, info_msg)
+        self.assertIsNotNone(match, "Snapshot creation message not found in info_msg.")
```
This fails if the system has any errors or warnings for any of the devices in the snapshot created by FAS. For example, on rocket:

```
ncn-m001:/mnt/developer/haasken/CRAYSAT-1957 # ./opt/cray/tests/install/csm-testing-python-venv/bin/test_sat_functional firmware.test_firmware.TestFirmware.test_firmware_creates_and_deletes_snapshot
test_firmware_creates_and_deletes_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that running `sat firmware` creates a snapshot and outputs a report ... FAIL

======================================================================
FAIL: test_firmware_creates_and_deletes_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that running `sat firmware` creates a snapshot and outputs a report
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/developer/haasken/CRAYSAT-1957/opt/cray/tests/install/csm-testing-python-venv/lib64/python3.6/site-packages/csm_testing/tests/sat_functional/firmware/test_firmware.py", line 60, in test_firmware_creates_and_deletes_snapshot
    self.assertIsNotNone(match, "Snapshot creation message not found in info_msg.")
AssertionError: unexpectedly None : Snapshot creation message not found in info_msg.

----------------------------------------------------------------------
Ran 1 test in 87.974s

FAILED (failures=1)
```

We should tolerate errors like this instead of failing. I would suggest updating your code to look for each info message independently. I would also suggest moving this to a helper method in this class because every test that creates a snapshot should probably capture the name of the snapshot that was created so it can delete it at the end of the test (in a `tearDown` method).

Here is what the errors look like on rocket, for more context:

```
ncn-m001:/mnt/developer/haasken/CRAYSAT-1957 # sat firmware
INFO: Snapshot SAT-2025-7-9-22-12-40 created successfully.
ERROR: Error getting firmware for x3000c0s11b0 target MB_CPLD1: GET https://x3000c0s11b0/redfish/v1/UpdateService/FirmwareInventory/MB_CPLD1 giving up after 1 attempt(s): Get "https://x3000c0s11b0/redfish/v1/UpdateService/FirmwareInventory/MB_CPLD1": context deadline exceeded
...
```

They seem to be not occurring every time I run it on that system. We should account for the possibility of errors like this which are outside of SAT's control.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+            self.assertIn(col, output)
+
+        # Now delete the snapshot to keep the system clean
+        delete_command = f"sat firmware --delete-snapshot {snapshot_name}"
```
This is good to test in a test case like this, but we also want to make sure all snapshots from other tests are cleaned up. To do so, add a `tearDown` method. That `tearDown` method could either:

* find and delete all SAT snapshots that have the expected prefix (perhaps between the start and end time of the tests executed in the class).
* or delete all SAT snapshots that have been explicitly marked for deletion. To do this, you'd need to mark them for deletion after they're created. That's probably a less robust solution.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+        list_command = "sat firmware --snapshots"
+        proc = subprocess.run(shlex.split(list_command), stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+        snapshots = [line.strip() for line in proc.stdout.decode().splitlines() if line.strip() and not line.startswith("#")]
+        self.assertGreater(len(snapshots), 0, "No firmware snapshots available for testing.")
```
I would recommend trying to make the test methods more self-contained, so create a snapshot in this test (or in the `setUpClass` method, so it always exists for all test methods in this class and can be re-used). Then use that snapshot in this test method.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+        list_command = "sat firmware --snapshots"
+        proc = subprocess.run(shlex.split(list_command), stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+        snapshots = [line.strip() for line in proc.stdout.decode().splitlines() if line.strip() and not line.startswith("#")]
+        self.assertGreaterEqual(len(snapshots), 2, "Less than two firmware snapshots available for testing.")
```
I would suggest making sure the test is self-contained by creating the needed snapshots in this test method or in the `setUpClass` method. You can make them small snapshots by using the `--xname` option to ensure good performance.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+
+    def test_firmware_query_nonexistent_snapshot(self):
+        """Test that `sat firmware --snapshots NON_EXISTENT_SNAPSHOT` querying a non-existent firmware snapshot."""
+        non_existent_snapshot = "NON_EXISTENT_SNAPSHOT"
```
Probably safe, but it might be better to generate a random and basically guaranteed to be unique value here (e.g. using the `uuid` module)

```
>>> import uuid
>>> str(uuid.uuid4())
'2713b54e-3687-44b1-8387-8f83fd85a5ab'
```

That's assuming that such a value is a valid FAS snapshot name, which I have not checked.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+        """Test that `sat firmware --xname-file <file>` prints firmware info for xnames listed in a file."""
+        xnames = get_xnames_from_firmware_output()
+
+        # Write the first two xnames to a file in the current directory
+        tmpfile_path = "test_xnames.txt"
+        with open(tmpfile_path, "w") as tmpfile:
+            tmpfile.write(f"{xnames[0]}\n{xnames[1]}\n")
```
This is probably not critical, but you can use the `tempfile` module to get unique temporary files.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+        pattern = r"^SAT-\d{4}-\d{1,2}-\d{1,2}-\d{1,2}-\d{1,2}-\d{1,2}$"
+        self.assertGreater(len(lines), 0, "No snapshots listed.")
+        for line in lines:
+            self.assertRegex(line, pattern, f"Line does not look like a snapshot name: {line}")
```
As discussed today, we should not try to ensure every snapshot matches this regular expression. If you have a `setUpClass` method that sets up a couple snapshots for use by the test methods, it would make sense to expect at least some of the snapshots match this naming convention.

You could even go further and expect to see the snapshots created by this test in the output of `sat firmware --snapshots`. The names of those snapshots would be saved as class vars (accessible through `self` like `self.first_snapshot_name`) in the `setUpClass` method.

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+        if fields and fields[0].startswith('x') and fields[0] not in xnames:
+            xnames.append(fields[0])
+    return xnames
+
```
PEP8 requires another blank line

```suggestion


```

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+        proc = subprocess.run(shlex.split(query_command), stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE, check=False)
+        stderr_output = proc.stderr.decode()
+        self.assertTrue(
```
Suggest using `assertIn`

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,188 @@
+            "WARNING: Snapshot NON_EXISTENT_SNAPSHOT does not exist." in stderr_output,
+            "Expected warning about non-existent snapshot not found."
+        )
+        self.assertTrue(
```
Suggest using `assertIn`

haasken-hpe reviewed (changes requested):
The test cases are providing sufficient coverage of the `sat firmware` command, so that looks good.

However, I think the performance of these tests could be significantly improved with some of my suggestions, and we also need to ensure that these tests will succeed across a variety of systems (including vshasta ultimately).

annapoorna-s-alt commented:
Testing on rocket after addressing the comments:

```
ncn-m001:/mnt/developer/sann/automation-test/firmware/firmware-2 # ./opt/cray/tests/install/csm-testing-python-venv/bin/test_sat_functional firmware.test_firmware
test_firmware_describe_existing_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots SNAPSHOT` describes an existing firmware snapshot. ... ok
test_firmware_list_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots` prints a list of snapshot names. ... ok
test_firmware_query_multiple_snapshots (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots SNAPSHOT_1 SNAPSHOT_2` prints the details for two snapshots. ... ok
test_firmware_query_multiple_xnames (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware -x <xname1>,<xname2>` prints firmware info for multiple xnames. ... ok
test_firmware_query_nonexistent_snapshot (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --snapshots NON_EXISTENT_SNAPSHOT` querying a non-existent firmware snapshot. ... ok
test_firmware_query_single_xname (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware -x <xname>` prints firmware info for a single xname. ... ok
test_firmware_query_xnames_from_file (csm_testing.tests.sat_functional.firmware.test_firmware.TestFirmware)
Test that `sat firmware --xname-file <file>` prints firmware info for xnames listed in a file. ... ok

----------------------------------------------------------------------
Ran 7 tests in 97.193s

OK
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,214 @@
+
+        # Ensure there are at least two xnames for testing
+        if len(cls.xnames) < 2:
+            raise RuntimeError("Not enough xnames available for testing.")
```
It would probably be better to skip the tests than to raise an exception. We wouldn't want failures on a system that has no xnames (e.g., perhaps vShasta?).

annapoorna-s-alt replied:
Addressed

haasken-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,214 @@
+    @classmethod
+    def tearDownClass(cls):
+        """Clean up all SAT snapshots created during the test class execution."""
+        cls.test_end_time = datetime.datetime.now()
```
It is a little odd that we are using a class attribute to store the end time of all the tests in the class. This would be set on the class object itself and visible across all instances of the class. I don't think it functionally matters though since I don't think anything creates multiple instances of the same class in a single Python interpreter.

I can't think of a better solution. What we'd ideally want is a method in `unittest.TestCase` that is run after all tests in the class are executed, but we'd want it to be an instance method instead of a class method, so it can see instance vars where we would store the start and end time. I don't think such a method exists in a TestCase, so this is probably fine.

annapoorna-s-alt replied:
Yeah, Even I don't think the better way. Hopefully it's fine.

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+import datetime
+import uuid
+import tempfile
+
+SAT_FIRMWARE_HEADER = ['xname', 'name', 'target_name', 'version']
+
+
+def get_xnames() -> list:
```
```suggestion
from typing import List

SAT_FIRMWARE_HEADER = ['xname', 'name', 'target_name', 'version']


def get_xnames() -> List[str]:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+class TestFirmware(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
```
```suggestion
    def setUpClass(cls) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+            xname_file.write(f"{cls.xnames[0]}\n{cls.xnames[1]}\n")
+
+    @classmethod
+    def tearDownClass(cls):
```
```suggestion
    def tearDownClass(cls) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        cls.temp_dir.cleanup()
+
+    @classmethod
+    def create_firmware_snapshot(cls, xname:str) -> str:
```
```suggestion
    def create_firmware_snapshot(cls, xname: str) -> str:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+            raise RuntimeError("Snapshot creation failed. Errors encountered:\n" + info_msg)
+
+    @classmethod
+    def delete_sat_snapshots_in_time_window(cls, start_time, end_time):
```
```suggestion
    def delete_sat_snapshots_in_time_window(
        cls, start_time: datetime.datetime, end_time: datetime.datetime
    ) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+                    cls.delete_snapshot(snapshot)
+
+    @staticmethod
+    def delete_snapshot(snapshot_name):
```
```suggestion
    def delete_snapshot(snapshot_name: str) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        subprocess.run(shlex.split(command), stdout=subprocess.PIPE,
+                       stderr=subprocess.PIPE, check=True )
+
+    def test_firmware_describe_existing_snapshot(self):
```
```suggestion
    def test_firmware_describe_existing_snapshot(self) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        for col in SAT_FIRMWARE_HEADER:
+            self.assertIn(col, output)
+
+    def test_firmware_query_multiple_snapshots(self):
```
```suggestion
    def test_firmware_query_multiple_snapshots(self) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        for col in SAT_FIRMWARE_HEADER:
+            self.assertIn(col, output)
+
+    def test_firmware_query_nonexistent_snapshot(self):
```
```suggestion
    def test_firmware_query_nonexistent_snapshot(self) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        error_message = 'ERROR: Failed to get snapshots: No firmware found.'
+        self.assertIn(error_message, stderr_output, f'Expected error not found in stderr: {stderr_output}')
+
+    def test_firmware_query_single_xname(self):
```
```suggestion
    def test_firmware_query_single_xname(self) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        for col in SAT_FIRMWARE_HEADER:
+            self.assertIn(col, xname_output)
+
+    def test_firmware_query_multiple_xnames(self):
```
```suggestion
    def test_firmware_query_multiple_xnames(self) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        for col in SAT_FIRMWARE_HEADER:
+            self.assertIn(col, multi_xname_output)
+
+    def test_firmware_query_xnames_from_file(self):
```
```suggestion
    def test_firmware_query_xnames_from_file(self) -> None:
```

mharding-hpe commented on src/csm_testing/tests/sat_functional/firmware/test_firmware.py:
```diff
@@ -0,0 +1,211 @@
+        for col in SAT_FIRMWARE_HEADER:
+            self.assertIn(col, file_xname_output)
+
+    def test_firmware_list_snapshots(self):
```
```suggestion
    def test_firmware_list_snapshots(self) -> None:
```

haasken-hpe commented:
@mharding-hpe, are type hints required in Python code in this repo? I noticed most of your changes had to do with fixing or adding those.

mharding-hpe commented:
> @mharding-hpe, are type hints required in Python code in this repo? I noticed most of your changes had to do with fixing or adding those.

@haasken-hpe Not *required* but strongly encouraged. We have a lot of old code that we still use that has no type hinting, but I'm hoping to eventually clean that up. I try to fix up incoming code as much as possible. 

I spent a lot of time fully type annotating the BOS code, and I recently helped the rack resiliency team get their code fully annotated. In both cases, we now run `mypy` automatically for builds and PRs, and in both repos it identified bugs that otherwise had been undetected. That convinced me that it's a good thing to aspire for in all of our codebases. But given where CSM is, and given that the old Python code prevents us from being able to use `mypy` yet, I can't in good conscience say that type hints are absolutely required.

haasken-hpe commented:
> @haasken-hpe Not _required_ but strongly encouraged. We have a lot of old code that we still use that has no type hinting, but I'm hoping to eventually clean that up. I try to fix up incoming code as much as possible.

Understood, thanks @mharding-hpe. I'll keep this in mind.
